### PR TITLE
WebGLRenderer: Made MultiMaterial more robust

### PR DESCRIPTION
**Pull request by Mugen87**
see https://github.com/mrdoob/three.js/pull/10636#issuecomment-276531885

This PR changes two things in `.projectObject()`:

1. If the user provides a `MultiMaterial` with less materials than expected, the user gets informed with a new warning. This avoids a runtime error mentioned in the linked PR.

2. If a geometry with no groups is used in combination with `MultiMaterial`, the method now creates a default group. This avoids no rendered objects like in this case: https://jsfiddle.net/oooao6qb/6/

### CODE REVIEW

**Comment by Mugen87** [comment]
The last commit implemented some feedback of @WestLangley . The code does not create a default group anymore. Instead it logs a warning when a `MultiMaterial` is used but groups are not specified.

File: src/renderers/WebGLRenderer.js [hunk 1]
```diff
@@ -1434,14 +1434,44 @@ function WebGLRenderer( parameters ) {
...
 
 								var group = groups[ i ];
 								var groupMaterial = materials[ group.materialIndex ];
 
-								if ( groupMaterial.visible === true ) {
+								if ( groupMaterial === undefined ) {
+
+									console.warn( 'THREE.WebGLRenderer: MultiMaterial has insufficient amount of materials for geometry. %i material(s) expected but only %i provided.', groups.length, materials.length );
+
+								} else {
```
Comment (mrdoob) [inline src/renderers/WebGLRenderer.js]: How about just...

```js
} else if ( groupMaterial.visible === true ) {

    pushRenderItem( object, geometry, groupMaterial, _vector3.z, group );

}
```

File: src/renderers/WebGLRenderer.js [hunk 2]
```diff
@@ -1434,17 +1434,35 @@ function WebGLRenderer( parameters ) {
...
 
-									pushRenderItem( object, geometry, groupMaterial, _vector3.z, group );
+									var group = groups[ i ];
+									var groupMaterial = materials[ group.materialIndex ];
+
+									if ( groupMaterial === undefined ) {
+
+										console.warn( 'THREE.WebGLRenderer: MultiMaterial has insufficient amount of materials for geometry. %i material(s) expected but only %i provided.', groups.length, materials.length );
+
+									} else {
```
Comment (mrdoob) [inline src/renderers/WebGLRenderer.js]: } else if ( groupMaterial.visible === true ) {

    pushRenderItem( object, geometry, groupMaterial, _vector3.z, group );

}

Comment (Mugen87) [inline src/renderers/WebGLRenderer.js]: Done

**Comment by WestLangley** [comment]
If we want to go this route then we need to be sure not to add groups unnecessarily. For example, in `WireframeGeometry`, we modify the original geometry by adding a group.

https://github.com/mrdoob/three.js/blob/master/src/geometries/WireframeGeometry.js#L85-L91

**Comment by Mugen87** [comment]
Maybe we can remove the mentioned group after generated the edge list via `.clearGroups()`. This will be the easiest way to keep the groups array untouched.

Should we do this in a separate PR?

**Comment by WestLangley** [comment]
Yes, a separate PR. Maybe this pattern will work:
```javascript
if ( groups.length === 0 ) {

	groups = [ { start: 0, count: indices.count, materialIndex: 0 } ];

}
```

**Comment by Mugen87** [comment]
That's great! 👍

**Comment by mrdoob** [comment]
Thanks!

**Pull request merged**